### PR TITLE
Cortx 29082 : Link and Unlink Bucket Operation

### PR DESCRIPTION
**Pull request by abhijit1patil**
# Problem Statement
-  [Link and Unlink Bucket](https://jts.seagate.com/browse/CORTX-29082)

# Design
-  [Link Bucket REST API Specification](https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/917570294/LINK+BUCKET+REST+API+Specifications)
-  [Unlink Bucket REST API Specification](https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/917570301/UNLINK+BUCKET+REST+API+Specifications)

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: csm/core/controllers/rgw/s3/base.py [hunk 1]
```diff
@@ -16,10 +16,25 @@
...
+    invalidate_empty_values : method with marshmallo schema validates_schema decorator
+    """
+    @classmethod
+    @validates_schema
+    def invalidate_empty_values(data, **kwargs):
+        """Method invalidates the empty strings."""
+        for key, value in data.items():
+            if value is not None and not str(value).strip():
+                raise ValidationError(f"{key}: Can not be empty")
+
```
Comment (rohitkolapkar) [inline csm/core/controllers/rgw/s3/base.py:35]: please check after the movement of this logic to parent class, are validations working for user APIs and Key APIs too.

File: csm/core/controllers/rgw/s3/bucket.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        super().__init__(request, const.RGW_S3_BUCKET_SERVICE)
+
+    @CsmAuth.permissions({Resource.S3_BUCKET: {Action.UPDATE}})
+    @Log.trace_method(Log.DEBUG)
+    async def put(self):
+        """
+        PUT REST implementation for  s3 bucket.
+        """
+        Log.debug(f"Handling s3 bucket PUT request"
+                  f" user_id: {self.request.session.credentials.user_id}")
```
Comment (rohitkolapkar) [inline csm/core/controllers/rgw/s3/bucket.py:79]: make this log as info

**Review by rohitkolapkar — CHANGES_REQUESTED** [review]

**Review by rohitkolapkar — CHANGES_REQUESTED** [review]
Import bucket controller in routes.py

File: csm/core/controllers/rgw/s3/bucket.py (hunk 2 above)
Comment (abhijit1patil) [reply csm/core/controllers/rgw/s3/bucket.py:79]: Done

File: csm/core/controllers/rgw/s3/base.py (hunk 1 above)
Comment (abhijit1patil) [reply csm/core/controllers/rgw/s3/base.py]: Done

**Pull request closed without merging**